docs: add a section about latency and how it's measured (#16734)
coder/coder: docs/admin/networking/index.md

In general, [port forwarded](./port-forwarding.md) web apps are faster than
dashboard-accessed web apps.

## Latency

Coder measures and reports several types of latency, providing insights into the performance of your deployment. Understanding these metrics can help you diagnose issues and optimize the user experience.

There are three main types of latency metrics for your Coder deployment:

- Dashboard-to-server latency:
  

>   The Coder UI measures round-trip time to the Coder server using the browser's Performance API.

Reviewer: Coder server or workspace proxy

Author: updated to

```diff
The Coder UI measures round-trip time to the Coder server
+ or workspace proxy
using the browser's Performance API.
```

[L208](https://github.com/coder/coder/pull/16734/files#diff-f2e36a38235181ed6696007214d293eda4ee5e9c5d8ecb9665b799beaeae8e54R208)

>   This appears in the user interface next to your username, showing how responsive the dashboard is.

Reviewer: It's a bit more complex if there are multiple workspace proxies. The dashboard is always served from the primary Coder server deployment, which will be one specific entry, although I don't think we do anything special to indicate which one, unfortunately.

The other workspace proxies serve applications from the workspace, like web IDEs, the web terminal, etc. A low ping there will help those applications be more responsive, but it only measures part of the round-trip. The proxy to workspace leg is not measured.

Author: I had trouble figuring out how to add that to this doc in a way that doesn't take it too far off course, so I added it to `workspace-proxies.md` in https://github.com/coder/coder/pull/16734/commits/9e7dc7b1d01e0c27c7cf1ff4cb331f13e747a119

I also force-pushed to this branch because I repointed it to `main` so that the latency info isn't tied to a rework of the networking docs in general. hopefully I did that as cleanly as possible 🤞

- Workspace connection latency:
  
  When users connect to workspaces, Coder measures and displays the latency between the user and:
  - Workspace (direct P2P connection when possible)
  - DERP relay servers (when P2P isn't possible)

>   This latency is visible in workspace cards and resource pages. It shows the round-trip time in milliseconds.

Reviewer: Ok, I've looked into this, and admittedly I'm a bit embarassed by what we built.

The latency on the Agent resource card is between the Agent and the shown relay. It never shows P2P (which isn't possible to achieve via the web browser anyway).

The estimated total latency if you were to access a workspace via the browser would be if you added the numbers by your username to the numbers you see in the Agent card, then that would give you a rough picture of the total latency thru various relays.

- **Green** (<150ms): Good performance.
- **Yellow** (150-300ms): Moderate latency that might affect user experience.
- **Red** (>300ms): High latency that will noticeably affect user experience.

### View latency information

- **Dashboard**: The global latency indicator appears in the top navigation bar.
- **Workspace list**: Each workspace shows its connection latency.
- **Health dashboard**: Administrators can view advanced metrics including database latency.
- **CLI**: Use `coder ping <workspace>` to measure and analyze latency from the command line.

### Factors that affect latency

- **Geographic distance**: Physical distance between users, Coder server, and workspaces.
- **Network connectivity**: Quality of internet connections and routing.
- **Infrastructure**: Cloud provider regions and network optimization.
- **P2P connectivity**: Whether direct connections can be established or relays are needed.

### How to optimize latency

To improve latency and user experience:

- **Deploy workspace proxies**: Place [proxies](./workspace-proxies.md) in regions closer to users.
- **Use P2P connections**: Ensure network configurations permit direct connections.

> - **Regional deployments**: Place Coder servers in regions where most users work.

Reviewer: Careful with the pluralization here. There should only be one Coder server deployment, located somewhere close to the most users. It can have multiple replicas for scale.

Author: https://github.com/coder/coder/pull/16734/files#diff-f2e36a38235181ed6696007214d293eda4ee5e9c5d8ecb9665b799beaeae8e54R250-R253

```markdown
- **Deploy workspace proxies**: Place [proxies](./workspace-proxies.md) in regions closer to users, connecting back to your single Coder server deployment.
- **Use P2P connections**: Ensure network configurations permit direct connections.
- **Strategic placement**: Deploy your Coder server in a region where most users work.
- **Network configuration**: Optimize routing between users and workspaces.
```

To learn more, see [Workspace Proxies](./workspace-proxies.md).

## Latency

Coder measures and reports several types of latency, providing insights into the performance of your deployment. Understanding these metrics can help you diagnose issues and optimize the user experience.

There are three main types of latency metrics for your Coder deployment:

- Dashboard-to-server latency:

  The Coder UI measures round-trip time to the Coder server or workspace proxy using built-in browser timing capabilities.

  This appears in the user interface next to your username, showing how responsive the dashboard is.

- Workspace connection latency:

  The latency shown on the workspace dashboard measures the round-trip time between the workspace agent and its DERP relay server.

  This metric is displayed in milliseconds on the workspace dashboard and specifically shows the agent-to-relay latency, not direct P2P connections.

>   Total end-to-end latency experienced by a user is a combination of dashboard-to-server latency plus this agent-to-relay latency.

Author: ```suggestion
  To estimate the total end-to-end latency experienced by a user, add the dashboard-to-server latency to this agent-to-relay latency.
```

- Database latency:

>   For administrators, Coder monitors and reports database query performance in the health dashboard.

Author: > Ok, I've looked into this, and admittedly I'm a bit embarassed by what we built.
> 
> The latency on the Agent resource card is between the Agent and the shown relay. It never shows P2P (which isn't possible to achieve via the web browser anyway).
> 
> The estimated total latency if you were to access a workspace via the browser would be if you added the numbers by your username to the numbers you see in the Agent card, then that would give you a rough picture of the total latency thru various relays.

@spikecurtis copying your comments over so they don't get lost

I removed the lines you had commented on, and replaced them with this. Do you think it helps to point out `not direct P2P connections` or should we cut that?
